marlowe-dashboard-client/src/Marlowe/Execution/Types.purs

```diff
@@ -1,60 +1,80 @@
 module Marlowe.Execution.Types
   ( ExecutionState
-  , PreviousState
+  , PreviousExecutionState
+  , CurrentExecutionState
   , PendingTimeouts
   , NamedAction(..)
   ) where
 
 import Prelude
 import Data.BigInteger (BigInteger)
+import Data.List (List)
 import Data.Map (Map)
 import Data.Maybe (Maybe)
 import Marlowe.Semantics (AccountId, Bound, ChoiceId, ChosenNum, Contract, Observation, Party, Payment, Slot, State, Token, TransactionInput, ValueId)
 
 type ExecutionState
-  = { previous :: Array PreviousState
-    , current ::
-        { state :: State
-        , contract :: Contract
-        }
+  = { previousExecutionStates :: Array PreviousExecutionState
+    , currentExecutionState :: CurrentExecutionState
     , mPendingTimeouts :: Maybe PendingTimeouts
     , mNextTimeout :: Maybe Slot
     }
 
--- This represents a previous step in the execution. The state property corresponds to the state before the
--- txInput was applied and it's saved as an early optimization to calculate the balances at each step.
-type PreviousState
-  = { txInput :: TransactionInput
+-- The current execution state is either the initial state of the contract, or a function of the
+-- last element in the `previous` array. It is saved as an early optimisation so that we don't need
+-- to recalculate payments and account balances all the the time.
+type CurrentExecutionState
+  = { contract :: Contract
     , state :: State
+    , payments :: List Payment
     }
 
--- This represents the timeouts that hasn't been applied to the contract. When a step of a contract has timeout
--- nothing happens until the next InputTransaction. This could be an IDeposit or IChoose for the continuation
--- contract, or an empty transaction to advance or even close the contract. We store a separate contract and state
--- than the "current" contract/state because this is the predicted state that we would be in if we applied an
--- empty transaction, and this allow us to extract the NamedActions that needs to be applied next.
--- Also, we store the timeouts as an Array because it is possible that multiple continuations have timeouted
--- before we advance the contract.
+-- Similar to the current execution state, the first three properties of the previous execution
+-- state either represent the initial state of the contract, or are a function of the previous 
+-- element in the `previous` array. The final property is the transaction input that was applied at
+-- that step. In other words, apply the transaction input to the contract and state here, and you
+-- get the *next* contract, state, and payments. The only thing we *need* to keep around,
+-- therefore, is the transaction input - but we remember the rest to save having to calculate it
+-- multiple times.
```

Comment:
similarly here, I'm not sure I understand this correctly.

What did you meant with this: `the first three properties of the previous execution
-- state either represent the initial state of the contract, or are a function of the previous 
-- element in the `previous` array`?